renderer.php

```diff
@@ -33,6 +33,7 @@
 class qtype_stack_renderer extends qtype_renderer {
 
     public function formulation_and_controls(question_attempt $qa, question_display_options $options) {
+        $qtype = new qtype_stack();
```

Comment:
It is definitely wrong to manually create a new instance of the qtype class. 

The comment on get_input_names_from_question_text says "This method is needed for validation, and to construct the editing form, and looking at what it does (using data_submitted) means that it is not safe to use when a student is attempting the question.

Mind you, this function is already been called from initialise_question_instance, so the comment is wrong, and calling from initialise_question_instance is probably theoretically wrong too.

I think the logic we want, in both initialise_question_instance and here, is just what  stack_utils::extract_placeholders($questiontext, 'input') does. 

Do we also need to worry about feedback placeholders getting messed up?

Also, there really should be at least one unit test added (probably copy one test function in question/type/stack/tests/walkthrough_adaptive_test.php, and add make a version that turns on forceclean. Make sure it fails without this bug fix.